lib/compiler-cranelift/src/compiler.rs

```diff
@@ -118,15 +121,20 @@ impl Compiler for CraneliftCompiler {
                 // if generate_debug_info {
                 //     context.func.collect_debug_info();
                 // }
+                let mut reader =
+                    MiddlewareBinaryReader::new_with_offset(input.data, input.module_offset);
+                reader.set_middleware_chain(
+                    self.config
+                        .middlewares
+                        .generate_function_middleware_chain(*i),
```

Comment:
I haven't seen the whole PR yet but just looking at this, I'm not quite sure what this does. If I had to guess, it's compiler codegen implemented as a middleware